roborock/devices/local_channel.py

```diff
@@ -39,18 +42,83 @@ class LocalChannel(Channel):
     format most parsing to higher-level components.
     """
 
-    def __init__(self, host: str, local_key: str):
+    def __init__(self, host: str, local_key: str, local_protocol_version: LocalProtocolVersion | None = None):
         self._host = host
         self._transport: asyncio.Transport | None = None
         self._protocol: _LocalProtocol | None = None
         self._subscribers: CallbackList[RoborockMessage] = CallbackList(_LOGGER)
         self._is_connected = False
-
-        self._decoder: Decoder = create_local_decoder(local_key)
-        self._encoder: Encoder = create_local_encoder(local_key)
+        self._local_key = local_key
+        self._local_protocol_version = local_protocol_version
+        self._connect_nonce = get_next_int(10000, 32767)
+        self._ack_nonce: int | None = None
+        self._update_encoder_decoder()
+
+    def _update_encoder_decoder(self):
+        self._encoder = create_local_encoder(
+            local_key=self._local_key, connect_nonce=self._connect_nonce, ack_nonce=self._ack_nonce
+        )
+        self._decoder = create_local_decoder(
+            local_key=self._local_key, connect_nonce=self._connect_nonce, ack_nonce=self._ack_nonce
+        )
         # Callback to decode messages and dispatch to subscribers
         self._data_received: Callable[[bytes], None] = decoder_callback(self._decoder, self._subscribers, _LOGGER)
 
+    async def _do_hello(self, local_protocol_version: LocalProtocolVersion) -> bool:
+        """Perform the initial handshaking."""
+        _LOGGER.debug(
+            "Attempting to use the %s protocol for client %s...",
+            local_protocol_version,
+            self._host,
+        )
+        request = RoborockMessage(
+            protocol=RoborockMessageProtocol.HELLO_REQUEST,
+            version=local_protocol_version.encode(),
+            random=self._connect_nonce,
+            seq=1,
+        )
+        try:
+            response = await self.send_message(
+                roborock_message=request,
+                request_id=request.seq,
+                response_protocol=RoborockMessageProtocol.HELLO_RESPONSE,
+            )
+            self._ack_nonce = response.random
+            self._local_protocol_version = local_protocol_version
+            self._update_encoder_decoder()
+
+            _LOGGER.debug(
+                "Client %s speaks the %s protocol.",
+                self._host,
+                local_protocol_version,
+            )
+            return True
+        except RoborockException as e:
+            _LOGGER.debug(
+                "Client %s did not respond or does not speak the %s protocol. %s",
+                self._host,
+                local_protocol_version,
+                e,
+            )
+            return False
+
+    async def hello(self):
```

Comment:
I think this can be a loop over protocols, sorted by whether or not they equal `self._local_protocol_version` if not none

e.g. 

```
    attempt_versions = [LocalProtocolVersion.V1, LocalProtocolVersion.L01]
    if self._local_protocol_version:
        attempt_versions.sort(key=lambda v: v == self._local_protocol_version). # Not sure if this works or needs to be reversed
    for version in attempt_versions:
        if await self._do_hello(version):
            return
    raise RoborockException("Failed to connect to device with any known protocol")
```